gh-89279: In ceval.c, redefine some macros for speed

gvanrossum opened the pull request:
In particular, `Py_DECREF`, `Py_XDECREF` and `Py_IS_TYPE` are redefined as macros that completely replace the inline functions of the same name. These three came out in the top four of functions that (in MSVC) somehow weren't inlined. (There was another, `_Py_atomic_load_32bit_impl`, but it is more difficult.

I got the top-N list using the recipe from [this bpo message](https://bugs.python.org/issue45116#msg406471).

<!-- issue-number: [gh-89279](https://github.com/python/cpython/issues/89279) -->
https://github.com/python/cpython/issues/89279
<!-- /issue-number -->

gvanrossum committed:
TEMPORARY hacks to show inline decisions

gvanrossum committed:
In ceval.c, redefine some macros for speed

In particular, Py_DECREF, Py_XDECREF and Py_IS_TYPE are redefined as macros
that completely replace the inline functions of the same name.
These three came out in the top four of functions that (in MSVC)
somehow weren't inlined.
(There was another, _Py_atomic_load_32bit_impl, but it is more difficult.)

vstinner commented on Python/ceval.c:
```diff
@@ -45,6 +45,20 @@
+// The MSVC compiler fails to inline these, and they're kind of important.
+
+#undef Py_DECREF
+#define Py_DECREF(arg) { PyObject *op = arg; if (--op->ob_refcnt == 0) { destructor d = Py_TYPE(op)->tp_dealloc; (*d)(op); } }
```
```suggestion
#define Py_DECREF(arg) do { PyObject *op = (arg); if (--op->ob_refcnt == 0) { destructor d = Py_TYPE(op)->tp_dealloc; (*d)(op); } } while (0)
```

vstinner commented on Python/ceval.c:
```diff
@@ -45,6 +45,20 @@
+#define Py_IS_TYPE(ob, type) ((PyObject *)(ob)->ob_type == (type))
+
+#undef Py_XDECREF
+#define Py_XDECREF(arg) { PyObject *op1 = arg; if (op1 != NULL) { Py_DECREF(op1); } }
```
```suggestion
#define Py_XDECREF(arg) do { PyObject *op = (arg); if (op != NULL) { Py_DECREF(op); } } while (0)
```

vstinner commented on Python/ceval.c:
```diff
@@ -45,6 +45,20 @@
 #endif
 
+#if !defined(Py_DEBUG)
+// The MSVC compiler fails to inline these, and they're kind of important.
```
```suggestion
// bpo-45116: The MSVC compiler fails to inline these in PGO build, and they're kind of important for performance.
```

vstinner commented on Python/ceval.c:
```diff
@@ -1579,6 +1593,7 @@ static inline bool
 is_method(PyObject **stack_pointer, int args) {
     return PEEK(args+2) != NULL;
 }
+#define is_method(stack_pointer, args) (PEEK((args)+2) != NULL)
```
Maybe remove the static inline function if the macro is always used, no?

vstinner reviewed (commented):
In C with well defined scopes, it's fine to reuse the same variable name in nested scopes ("op" in this case). It's just more annoying to debug.

C code:

```
#include <stdio.h>

int main()
{
    int x = 1;
    {
        int x = 2;
        {
            int x = 3;
            printf("x = %i\n", x);
        }
        printf("x = %i\n", x);
    }
    printf("x = %i\n", x);
    return 0;
}
```

Output:

```
x = 3
x = 2
x = 1
```

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into inline

gvanrossum committed:
Victor's review

gvanrossum committed:
Merge branch 'python:main' into inline

gvanrossum committed:
Add _Py_atomic_load_32bit_impl and use cast macro

gvanrossum committed:
Move _Py_atomic_load_32bit_impl out of '#if'

So that the assert() in the macro will be checked in debug mode.

vstinner reviewed (commented):
The Python/ceval.c changes LGTM, but I cannot review the 2 other modified files.

vstinner commented on Python/ceval.c:
```diff
@@ -44,6 +44,23 @@
 
+#if !defined(Py_DEBUG)
+// bpo-45116: The MSVC compiler fails to inline these in PGO build,
+// and they're kind of important for performance.
```
You might explain that MSVC can inline them, but it's just an arbitrary limit which cannot be configured.

neonene wrote "PR25244 told me the amount of code in _PyEval_EvalFrameDefault() is over the limit of PGO."

I suggest something like:

```suggestion
// bpo-45116: The MSVC compiler does not inline these 3 static inline functions in PGO build
// in _PyEval_EvalFrameDefault(), because this function is over the limit of PGO,
// and these limits which cannot be configured. Define them as macro to make sure that
// there are always inlined by the preprocessor.
```

Maybe tomorrow, MSVC PGO limits will change or become configurable, and it will become possible to remove these macros on more recent MSVC versions.

sweeneyde commented on Python/ceval.c:
```diff
@@ -44,6 +44,23 @@
+
+#endif
+
+#define _Py_atomic_load_32bit_impl(value, order) (assert((order) == _Py_memory_order_relaxed), *(value))
```
I notice the original signature is this:

```c
inline int _Py_atomic_load_32bit_impl(volatile int* value, int order)
```

Don't we need some sort of cast to `volatile int` so there's no risk of strange compiler optimizations?

sweeneyde replied:
Never mind, it looks okay: things are declared as `_Py_atomic_int` which uses `volatile int _value;`, so everything should be okay.

gvanrossum replied:
I do wonder why we are going through the hassle of these macros. Either @vstinner or @markshannon, do you remember what the concern here was? Why not just declare these are `volatile int32` (or `_Py_atomic_int`) and be done? All the callsites in this file specify "relaxed" mode. Are there platforms where things must be done differently?

sweeneyde replied:
Some answers here:
https://stackoverflow.com/questions/6627596/is-volatile-int-in-c-as-good-as-stdatomicint-of-c0x

Apparently, volatile variables are not required to be atomic.

gvanrossum replied:
I'm sorry, I wasn't arguing for `volatile` instead of `_Py_atomic_int`. But since, with the `_Py_atomic_int` declaration and the `_Py_memory_order_relaxed` memory order, apparently the whole `_Py_atomic_load_relaxed(x)` call degenerates to `*x`, why have a macro or inline function at all?

gvanrossum replied:
The more I look at pycore_atomic.h the more I fail to believe that anybody knows what it does. And even more so for dynamic_annotations.h. :-(

Anyway, I should probably override just `_Py_atomic_load_relaxed`, not `_Py_atomic_load_32bit_impl`. The former is more specific. And I should move it back inside the `#if !defined(Py_DEBUG)` block. (Hm, is `#ifndef` out of style?)

gvanrossum committed:
Go back to full PGO test suite

gvanrossum committed:
Longer comment; improve _Py_atomic_load_relaxed

gvanrossum committed:
Only specialize atomic-load-relaxed if MSC

gvanrossum committed:
Try _Py_atomic_load_relaxed_int32

gvanrossum committed:
MSC is better off with the default Py_UNREACHABLE()

gvanrossum committed:
Update bpo numbers to GH issue numbers

gvanrossum committed:
Split accidentally merged lines

gvanrossum committed:
Delete temporary /d2inlinelogfull option

gvanrossum committed:
Merge branch 'main' into inline

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

vstinner commented on Python/ceval.c:
```diff
@@ -44,6 +44,32 @@
+
+// GH-89279: Similar to above, force inlining by using a macro.
+#if defined(_MSC_VER) && SIZEOF_INT == 4
+#define _Py_atomic_load_relaxed_int32(ATOMIC_VAL) (assert(sizeof((ATOMIC_VAL)->_value) == 4), *((volatile int*)&((ATOMIC_VAL)->_value)))
```
You may use static_assert() to avoid adding a runtime check:

```suggestion
#define _Py_atomic_load_relaxed_int32(ATOMIC_VAL) \
    (static_assert(sizeof((ATOMIC_VAL)->_value) == 4, "expect 32-bit int"), \
     *((volatile int*)&((ATOMIC_VAL)->_value)))
```

vstinner commented on Python/ceval.c:
```diff
@@ -44,6 +44,32 @@
 #  error "ceval.c must be build with Py_BUILD_CORE define for best performance"
 #endif
 
+#ifndef Py_DEBUG
```
if only MSVC is affected, only redefine these macros for MSVC, no?

```suggestion
#if defined(_MSC_VER) && !defined(Py_DEBUG)
```

gvanrossum replied:
I prefer us running the same code on all platforms if possible, so that it's less likely that someone accidentally breaks the Windows version by want appears to be an innocent change on Linux.

(I have to make an exception for load_relaxed because it is heavily platform-specific.)

vstinner replied:
To be extra safe, you may also check Py_REF_DEBUG, even if Py_REF_DEBUG *should not* be defined if Py_DEBUG is defined.

```suggestion
#if !defined(Py_DEBUG) && !defined(Py_REF_DEBUG)
```

Or maybe object.h should fail with ``#error`` with Py_REF_DEBUG is defined whereas Py_DEBUG is not.

vstinner commented on Python/ceval.c:
```diff
@@ -1571,10 +1597,7 @@ typedef struct {
-is_method(PyObject **stack_pointer, int args) {
-    return PEEK(args+2) != NULL;
-}
+#define is_method(stack_pointer, args) (PEEK((args)+2) != NULL)
```
If it must be a macro to avoid reintroducing gh-89279, maybe add a comment to explain that?

vstinner commented on Include/pymacro.h:
```diff
@@ -122,8 +122,6 @@
 #elif defined(__clang__) || defined(__INTEL_COMPILER)
 #  define Py_UNREACHABLE() __builtin_unreachable()
-#elif defined(_MSC_VER)
-#  define Py_UNREACHABLE() __assume(0)
```
Is this change related? If it's only an issue in ceval.c, maybe redefine the macro there?

neonene replied:
As a fix only for x64, MSVC can use default case (`unknown opcode`) instead of current `TARGET(CACHE)` branch.

neonene replied:
`__assume(0)` in `TARGET(CACHE)` seems not to cause slowdown with this PR on x64. I'm fine with leaving as it is to avoid adding a confusing comment.

gvanrossum replied:
Yes, this was related -- @neonene [claims ](https://github.com/faster-cpython/ideas/issues/321#issuecomment-1075315759)that code using `__assume(0)` performs worse than the default (see other thread). But I'm not convinced that it is worth it and his latest comments seem to indicate that it's only a problem on x86 (i.e., 32-bit), not on x64, so I think I'll take it out.

gvanrossum committed:
Revert "MSC is better off with the default Py_UNREACHABLE()"

The evidence that `__assume(0)` is harmful is unclear and at best
seems to only apply to the x86 platform (i.e., 32-bit), which we
care about less and less.  So revert this change.

This reverts commit 5617fcdf08dda8ce87a81f8d8b190715091b4564.

gvanrossum committed:
Use static_assert() in _Py_atomic_load_relaxed_int32

Co-authored-by: Victor Stinner <vstinner@python.org>

gvanrossum committed:
Add comment referring to GH-89279 for is_method()

gvanrossum committed:
static_assert() is only a C++ feature

In GCC there's _Static_assert(), but MSVC only has _STATIC_ASSERT
and that requires crtdbg.h. I don't think it's worth the hassle.

Revert "Use static_assert() in _Py_atomic_load_relaxed_int32"

This reverts commit 9a5c57caee78c31ed99cc42e99bfe1d4b749498c.

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into inline

vstinner commented on Python/ceval.c:
```diff
@@ -44,6 +44,32 @@
+// preprocessor.
+
+#undef Py_DECREF
+#define Py_DECREF(arg) do { PyObject *op = _PyObject_CAST(arg); if (--op->ob_refcnt == 0) { destructor d = Py_TYPE(op)->tp_dealloc; (*d)(op); } } while (0)
```
Such long line is one of the reason why I wrote PEP 670 :-) I propose to make the code more readable by putting one statement per line, as done in static inline functions:

```c
#ifndef Py_DEBUG
// GH-89279: The MSVC compiler does not inline these static inline functions
// in PGO build in _PyEval_EvalFrameDefault(), because this function is over
// the limit of PGO, and that limit cannot be configured.
// Define them as macros to make sure that they are always inlined by the
// preprocessor.

#undef Py_DECREF
#define Py_DECREF(arg) \
    do { \
        PyObject *op = _PyObject_CAST(arg); \
        if (--op->ob_refcnt == 0) { \
            destructor dealloc = Py_TYPE(op)->tp_dealloc; \
            (*dealloc)(op); \
        } \
    } while (0)

#undef Py_XDECREF
#define Py_XDECREF(arg) \
    do { \
        PyObject *xop = _PyObject_CAST(arg); \
        if (xop != NULL) { \
            Py_DECREF(xop); \
        } \
    } while (0)

#undef Py_IS_TYPE
#define Py_IS_TYPE(ob, type) \
    (_PyObject_CAST_CONST(ob)->ob_type == (type))

#endif

// GH-89279: Similar to above, force inlining by using a macro.
#if defined(_MSC_VER) && SIZEOF_INT == 4
#define _Py_atomic_load_relaxed_int32(ATOMIC_VAL) \
    (assert(sizeof((ATOMIC_VAL)->_value) == 4), \
     *((volatile int*)&((ATOMIC_VAL)->_value)))
#else
#define _Py_atomic_load_relaxed_int32(ATOMIC_VAL) \
    _Py_atomic_load_relaxed(ATOMIC_VAL)
#endif
```

I declared Py_XDECREF just after Py_DECREF. I also renamed "op1" to "xop" in Py_XDECREF.

**UPDATE:** Oh, and to copy/paste Py_IS_TYPE(), I used _PyObject_CAST_CONST().

vstinner reviewed (commented):
I would prefer to add ``#if defined(_MSC_VER)``, but I'm fine with re-declare the 3 macros on all platforms.

gvanrossum committed:
Also redefine _Py_DECREF_SPECIALIZED as a macro

vstinner commented on Python/ceval.c:
```diff
@@ -44,6 +44,34 @@
+
+#undef _Py_DECREF_SPECIALIZED
+#define _Py_DECREF_SPECIALIZED(arg, dealloc) do { PyObject *op = _PyObject_CAST(arg); if (--op->ob_refcnt == 0) { (dealloc)(op); } } while (0)
+#endif
```
```suggestion
#ifndef Py_DEBUG
// GH-89279: The MSVC compiler does not inline these static inline functions
// in PGO build in _PyEval_EvalFrameDefault(), because this function is over
// the limit of PGO, and that limit cannot be configured.
// Define them as macros to make sure that they are always inlined by the
// preprocessor.

#undef Py_DECREF
#define Py_DECREF(arg) \
    do { \
        PyObject *op = _PyObject_CAST(arg); \
        if (--op->ob_refcnt == 0) { \
            destructor dealloc = Py_TYPE(op)->tp_dealloc; \
            (*dealloc)(op); \
        } \
    } while (0)

#undef Py_XDECREF
#define Py_XDECREF(arg) \
    do { \
        PyObject *xop = _PyObject_CAST(arg); \
        if (xop != NULL) { \
            Py_DECREF(xop); \
        } \
    } while (0)

#undef Py_IS_TYPE
#define Py_IS_TYPE(ob, type) \
    (_PyObject_CAST_CONST(ob)->ob_type == (type))

#undef _Py_DECREF_SPECIALIZED
#define _Py_DECREF_SPECIALIZED(arg, dealloc) \
    do { \
        PyObject *op = _PyObject_CAST(arg); \
        if (--op->ob_refcnt == 0) { \
            (dealloc)(op); \
        } \
    } while (0)
#endif
```

vstinner reviewed (commented):
My suggestion to reformat macro on mutliple lines, now as a GitHub suggestion (and with _Py_DECREF_SPECIALIZED).

gvanrossum committed:
Reformat macros for readability

Co-authored-by: Victor Stinner <vstinner@python.org>

vstinner reviewed (approved):
LGTM, even if I didn't understand everything. But I trust the ones who ran extensive tests on Python benchmarks with different compiler options.

There are many compiler warnings. Are they related to the change?

> GitHub Actions / Address sanitizer
> Python/ceval.c#L2408
> function called through a non-compatible type

sweeneyde commented on Python/ceval.c:
```diff
@@ -44,6 +44,54 @@
+        if (--op->ob_refcnt == 0) { \
+            (dealloc)(op); \
+        } \
+    } while (0)
```
I don't have an asan build handy so I'm not sure, but this might silence some warnings.
```suggestion
    do { \
        PyObject *op = _PyObject_CAST(arg); \
        if (--op->ob_refcnt == 0) { \
            destructor d = (destructor)(dealloc);
            d(op); \
        } \
    } while (0)
```

gvanrossum committed:
Try to silence warnings

Co-authored-by: Dennis Sweeney <36520290+sweeneyde@users.noreply.github.com>

vstinner commented on Python/ceval.c:
```diff
@@ -44,6 +44,55 @@
+
+#undef Py_IS_TYPE
+#define Py_IS_TYPE(ob, type) \
+    (_PyObject_CAST_CONST(ob)->ob_type == (type))
```
I just removed _PyObject_CAST_CONST:

```suggestion
    (_PyObject_CAST(ob)->ob_type == (type))
```

vstinner commented on Python/ceval.c:
```diff
@@ -44,6 +44,55 @@
+    do { \
+        PyObject *op = _PyObject_CAST(arg); \
+        if (--op->ob_refcnt == 0) { \
+            destructor d = (destructor)(dealloc);
```
Hey, it's hard to write macro (hello PEP 670, haha):

```suggestion
            destructor d = (destructor)(dealloc); \
```

vstinner committed:
Update Python/ceval.c

vstinner committed:
Update Python/ceval.c

vstinner reviewed (approved):
LGTM.
